features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/OpennmsKafkaProducer.java

```diff
@@ -391,31 +374,41 @@ private void sendRecord(Callable<ProducerRecord<byte[], byte[]>> callable, Consu
 
         // Any offer that fails due to capacity overflow will simply be dropped and will have to wait until the next
         // sync to be processed so this is just a best effort attempt
-        if (!kafkaSendQueue.offer(new KafkaRecord(record, callback))) {
+        if (!kafkaSendQueue.offer(new KafkaRecord(record, forwardedRecord))) {
             RATE_LIMITED_LOGGER.warn("Dropped a Kafka record due to queue capacity being full.");
         }
     }
 
     private void processKafkaSendQueue() {
         //noinspection InfiniteLoopStatement
+        KafkaRecord kafkaRecord = null;
+        ProducerRecord<byte[], byte[]> producerRecord = null;
+        AtomicBoolean resendRecord = new AtomicBoolean(false);
         while (true) {
             try {
-                KafkaRecord kafkaRecord = kafkaSendQueue.take();
-                ProducerRecord<byte[], byte[]> producerRecord = kafkaRecord.getProducerRecord();
-                Consumer<RecordMetadata> consumer = kafkaRecord.getConsumer();
-
+                if (!resendRecord.get()) {
+                    kafkaRecord = kafkaSendQueue.take();
+                    producerRecord = kafkaRecord.getProducerRecord();
+                } else if (kafkaRecord == null) {
+                    return;
+                }
                 try {
-                    producer.send(producerRecord, (recordMetadata, e) -> {
-                        if (e != null) {
-                            LOG.warn("Failed to send record to producer: {}.", producerRecord, e);
-                            return;
-                        }
-                        if (consumer != null) {
-                            consumer.accept(recordMetadata);
-                        }
-                    });
+                    Future<RecordMetadata> future = producer.send(producerRecord);
+                    future.get();
```

Comment:
This is going to kill performance.

Reply:
I think there is really no way to keep the performance in async mode and maintain the order.  
May be other ways to mitigate this are having large timeout and having more retries.

Reply:
Do we need to keep order? That part isn't clear to me based on the previous comments.

Reply:
I think we need to keep the order at least for Alarms. This change would allow it but would impact performance since we are blocking for every message.

Reply:
If we need to enforce orderly dispatch from the queue, then yeah we have no choice but to wait until we know the  message didn't timeout (in the case we are doing retry on timeout). That will kill performance as discussed.

Is it sufficient to ignore ordering for timeouts and just send them ASAP instead? We could do that by putting the message at the front of the queue rather than the back. In this case we have best effort ordering while maintaining the original performance characteristics.

Reply:
I fear best effort may not be sufficient. Given that Alarms are sent with with `reductionKey` as key,  reversing the order for null alarm ( deleted) / regular alarm may be bad for maintaining correct state at consumer side.

Reply:
Yep, thats a good point. I don't think we can have our cake and eat it too in this case.

We can try other strategies, but I can't think of one off top of my head that is optimal.

Since we get a callback on ack, we can keep a separate data structure of ordered messages in flight and remove them on ack. With this, we can now track state while sending and upon receiving a timeout we can switch to sending sync. Now we know which in flight messages were unack'd and could attempt to wait until we flush those all out and receive acks before we transition back into the original async state. This is an attempt to keep us running async in steady state but switching to sync in failure scenario. I don't think it can guarantee ordering for all messages though.

I think this comes with additional problems, dealing with transient timeouts would be a huge pain. You would still fail to guarantee ordering in the case you sent `A` -> `B` -> `C` where `B` timed out but `A` and `C` did not, since you wouldn't know about B's timeout right away.

---

Do we care about the performance of emptying this queue? Maybe it is ok to do it sync?

Reply:
Yeah, I m thinking may be just for Alarms, we can do sync ( that actually allows it more linear delivery) if we think Alarms may not be that high volume.

Reply:
Chatted with @j-white on this. Since nodes/alarms will be consistent after a while since they will be synced from DB again even if they are timed out.

Only for events, we can re-insert messages that are timed out. I will make that change.

Reply:
@mattixtech Can you review this again. When the message times out, we add element to the beginning of queue just for events.

Reply:
Will take a look